bridges/snowbridge/pallets/system-v2/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		LocationConversionFailed,
+		AgentAlreadyCreated,
+		NoAgent,
+		UnsupportedLocationVersion,
+		InvalidLocation,
+		Send(SendError),
+	}
+
+	/// The set of registered agents
+	#[pallet::storage]
+	#[pallet::getter(fn agents)]
+	pub type Agents<T: Config> = StorageMap<_, Twox64Concat, AgentId, (), OptionQuery>;
+
+	/// Lookup table for foreign token ID to native location relative to ethereum
+	#[pallet::storage]
+	pub type ForeignToNativeId<T: Config> =
+		StorageMap<_, Blake2_128Concat, TokenId, xcm::v5::Location, OptionQuery>;
+
+	/// Lookup table for native location relative to ethereum to foreign token ID
+	#[pallet::storage]
+	pub type NativeToForeignId<T: Config> =
+		StorageMap<_, Blake2_128Concat, xcm::v5::Location, TokenId, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Sends a command to the Gateway contract to instantiate a new agent contract representing
+		/// `origin`.
+		///
+		/// Fee required: Yes
+		///
+		/// - `origin`: Must be `Location` of a sibling parachain
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::create_agent())]
+		pub fn create_agent(
+			origin: OriginFor<T>,
+			location: Box<VersionedLocation>,
+			fee: u128,
+		) -> DispatchResult {
+			let _ = match origin.clone().into() {
+				Ok(RawOrigin::Root) => Ok(Here.into()),
+				_ => T::SiblingOrigin::ensure_origin(origin),
+			}?;
+
+			let origin_location: Location =
+				(*location).try_into().map_err(|_| Error::<T>::UnsupportedLocationVersion)?;
+
+			let ethereum_location = T::EthereumLocation::get();
+			// reanchor to Ethereum context
+			let location = origin_location
+				.clone()
+				.reanchored(&ethereum_location, &T::UniversalLocation::get())
+				.map_err(|_| Error::<T>::LocationConversionFailed)?;
+
+			let agent_id = agent_id_of::<T>(&location)?;
+
+			// Record the agent id or fail if it has already been created
+			ensure!(!Agents::<T>::contains_key(agent_id), Error::<T>::AgentAlreadyCreated);
+			Agents::<T>::insert(agent_id, ());
+
+			let command = Command::CreateAgent {};
+
+			Self::send(origin_location.clone(), command, fee)?;
+
+			Self::deposit_event(Event::<T>::CreateAgent {
+				location: Box::new(origin_location),
+				agent_id,
+			});
+			Ok(())
+		}
+
+		/// Registers a Polkadot-native token as a wrapped ERC20 token on Ethereum.
+		/// Privileged. Can only be called by root.
+		///
+		/// Fee required: No
+		///
+		/// - `origin`: Must be root
+		/// - `location`: Location of the asset (relative to this chain)
+		/// - `metadata`: Metadata to include in the instantiated ERC20 contract on Ethereum
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::register_token())]
+		pub fn register_token(
+			origin: OriginFor<T>,
+			asset_id: Box<VersionedLocation>,
+			metadata: AssetMetadata,
+			fee: u128,
+		) -> DispatchResult {
+			let origin_location = match origin.clone().into() {
+				Ok(RawOrigin::Root) => Ok(Here.into()),
```

Comment:
@yrong why not just force root governance to go through AH system-frontend, to reduce complexity here?